File: abc181-200/abc196/d/main.py
H,W,A,B = map(int,input().split())

filled = [[False] * W for _ in range(H)]

def dfs(h,w,a,b):
	global filled, H, W, A, B
	if a > A or b > B:
		return 0
	if w == W:
		return dfs(h+1,0,a,b)
	if h == H:
		return 1
	if filled[h][w]:
		return dfs(h,w+1,a,b)
	# square
	square = dfs(h, w+1, a, b+1)
	# long tate
	tate = 0
	if h+1 < H:
		filled[h+1][w] = True
		tate = dfs(h, w+1, a+1, b)
		filled[h+1][w] = False
	# long yoko
	yoko = 0
	if w+1 < W:
		yoko = dfs(h, w+2, a+1, b)
	return square + tate + yoko

print(dfs(0,0,0,0))

File: abc181-200/abc196/d/main/src/main.rs
fn main() {
	proconio::input! {
		h: usize,
		w: usize,
		a: usize,
		b: usize
	}
	let filled: u32 = 0;
	println!("{}", dfs(0, 0, 0, 0, h, w, a, b, filled));
}

fn dfs(i: usize, j: usize, a: usize, b: usize, _h: usize, _w: usize, _a: usize, _b: usize, filled: u32) -> usize {
	if a > _a || b > _b {
		return 0;
	} else if i == _h {
		return 1;
	} else if j == _w {
		return dfs(i + 1, 0, a, b, _h, _w, _a, _b, filled);
	} else if filled >> (i * _w + j) & 1 != 0 {
		return dfs(i, j + 1, a, b, _h, _w, _a, _b, filled);
	}

	let mut tate: usize = 0;
	let mut yoko: usize = 0;
	let square: usize = dfs(i, j + 1, a, b + 1, _h, _w, _a, _b, filled);

	if i + 1 < _h {
		tate = dfs(i, j + 1, a + 1, b, _h, _w, _a, _b, filled | 1 << ((i + 1) * _w + j));
	}
	if j + 1 < _w {
		yoko = dfs(i, j + 2, a + 1, b, _h, _w, _a, _b, filled);
	}

	return tate + yoko + square;
}
